sdk/src/Core/RegionEndpoint/RegionEndpointProviderV3.cs

```diff
@@ -364,34 +374,46 @@ public void Add(string serviceName, RegionEndpoint.Endpoint endpoint, HashSet<st
             {
                 if (variants == null || variants.Count == 0)
                 {
-                    _serviceMap.Add(serviceName, endpoint);
+                    lock (_serviceMapLock)
+                    {
+                        _serviceMap.Add(serviceName, endpoint);
+                    }
                 }
                 else
                 {
-                    if (!_variantMap.ContainsKey(serviceName) || _variantMap[serviceName] == null)
+                    lock (_variantMapLock)
                     {
-                        _variantMap.Add(serviceName, new Dictionary<HashSet<string>, RegionEndpoint.Endpoint>(HashSet<string>.CreateSetComparer()));
+                        if (!_variantMap.ContainsKey(serviceName) || _variantMap[serviceName] == null)
+                        {
+                            _variantMap.Add(serviceName, new Dictionary<HashSet<string>, RegionEndpoint.Endpoint>(HashSet<string>.CreateSetComparer()));
+                        }
+                        _variantMap[serviceName].Add(variants, endpoint);
                     }
-                    _variantMap[serviceName].Add(variants, endpoint);
                 }
             }
 
             public bool TryGetEndpoint(string serviceName, HashSet<string> variants, out RegionEndpoint.Endpoint endpoint)
             {
                 if (variants == null || variants.Count == 0)
                 {
-                    return _serviceMap.TryGetValue(serviceName, out endpoint);
+                    lock (_serviceMapLock)
+                    {
+                        return _serviceMap.TryGetValue(serviceName, out endpoint);
+                    }
                 }
                 else
                 {
-                    if (!_variantMap.ContainsKey(serviceName))
-                    {
-                        endpoint = default(RegionEndpoint.Endpoint);
-                        return false;
-                    }
-                    else
+                    lock (_variantMapLock)
                     {
-                        return _variantMap[serviceName].TryGetValue(variants, out endpoint);
+                        if (!_variantMap.ContainsKey(serviceName))
+                        {
+                            endpoint = default(RegionEndpoint.Endpoint);
+                            return false;
```

Comment:
Verified its ok to `return ` inside of `lock` statement:
https://stackoverflow.com/questions/266681/should-a-return-statement-be-inside-or-outside-a-lock